File: Array | Dictionary | Set/group_anagram/py/group_anagram.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hashmap = defaultdict(list)
        for str in strs:
            keys = [0] * 26
            for char in str:
                keys[ord(char) - ord('a')] += 1
            
            hashmap[tuple(keys)].append(str)

        return list(hashmap.values())

sol = Solution()

sol.groupAnagrams(["eat","tea","tan","ate","nat","bat"])

File: Array | Dictionary | Set/group_anagram/js/group_anagram.js
var groupAnagrams = function(strs) {
    const hasmap = new Map();
    const result = [];

    for ( let str of strs ) {
        const keys = new Array(26).fill(0);

        for ( let char of str ) {
            keys[char.charCodeAt(0) - 'a'.charCodeAt(0)]++;
        }
        
        const keyString = keys.join('-');

        if ( hasmap.has(keyString) ) {
            hasmap.get(keyString).push(str)
        } else {
            hasmap.set( keyString, [str] );
        }
    }

    for ( let value of hasmap.values() ) {
        result.push(value)
    }

    return result;
};


console.log(groupAnagrams(["eat","tea","tan","ate","nat","bat"]));